### PR TITLE
Add panel for controlling stepper motor

### LINKED ISSUE #32: Add panel for stepper motor controller
The panel will both show the status of the stepper motor controller as well as allowing the user to set the position of the view mirror attached to the stepper motor.

See diagram in #24.

From the proposal:

> Upper left red box: Shows the status of the stepper motor controller,
the boxes marked zenith, nadir, home... allow the user to point and click the required position of the view mirror attached to the stepper motor. The blue highlight (in this case over cold_BB) indicates the current position of the mirror.

Aha! Link: https://imperial-college-ict.aha.io/requirements/RSE-122-21

### PR DESCRIPTION
**Pull request by alexdewar**
Closes #32.

This PR adds the panel for controlling the stepper motor (which is in the top left of the GUI for the old program). Users can click on one of the buttons to move to a preset angle specified by the name on the button or can go to a specific angle/step by choosing a number and clicking "GOTO". The background of the last-clicked button changes to blue, as was done in the old program.

In the process, I created a `DummyStepperMotor` class to support the functionality of the GUI. I know we said we wouldn't venture into making dummy devices in this sprint, but this seemed like the cleanest way to implement the GUI functionality. (Its implementation will no doubt change when we realise the actual stepper motor behaves a bit differently.) I put it in a separate `hardware` module and it communicates with the GUI using PyPubSub messages. This should keep the concerns neatly separated.

### CODE REVIEW

File: finesse/gui/stepper_motor_view.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+        self.setLayout(layout)
+
+    def _create_stepper_button(self, name: str) -> QPushButton:
+        """Create a button to move the motor to a preset position.
+
+        Args:
+            name: The name of the preset (will be converted to lowercase)
+        """
+        btn = QPushButton(name)
+        btn.clicked.connect(lambda: self._preset_clicked(btn))  # type: ignore
```
Comment (dalonsoa) [inline finesse/gui/stepper_motor_view.py:45]: Use `partial` from `functools`, as I did. `mypy` seems to be happy with it.

File: finesse/gui/stepper_motor_view.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        """Change the currently selected button, indicated by colour.
+
+        Args:
+            btn: The just-clicked button
+        """
+        if self.last_clicked:
+            self.last_clicked.setStyleSheet("")
+        btn.setStyleSheet("background-color: blue")
+        self.last_clicked = btn
+
```
Comment (dalonsoa) [inline finesse/gui/stepper_motor_view.py:58]: I'm not totally sure, but the way you're using this makes me think on button groups. Have a look at this in case that is what you actually want to use: https://doc.qt.io/qtforpython/PySide6/QtWidgets/QButtonGroup.html

File: finesse/gui/stepper_motor_view.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    def _preset_clicked(self, btn: QPushButton) -> None:
+        """Move the stepper motor to preset position."""
+        self._set_selected_button(btn)
+        pub.sendMessage("stepper.move", target=btn.text().lower())
+
+    def _goto_clicked(self) -> None:
```
Comment (dalonsoa) [inline finesse/gui/stepper_motor_view.py:64]: GOTO makes me think on the old Fortran 77...

File: finesse/hardware/dummy_stepper_motor.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+        Args:
+            target: The target angle (in degrees) or the name of a preset
+        """
+        if isinstance(target, str):
+            self.current_step = DummyStepperMotor.get_preset_step(target)
+        else:
+            step = round(self.steps_per_rotation * target / 360.0)
+            if step < 0 or step >= self.steps_per_rotation:
+                raise ValueError("step ")
```
Comment (dalonsoa) [inline finesse/hardware/dummy_stepper_motor.py:49]: So we need something more informative here?

File: tests/test_stepper_motor.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
+"""Tests for DummyStepperMotor."""
+import pytest
+
+from finesse.hardware.dummy_stepper_motor import DummyStepperMotor
+
+PRESETS = ("zenith", "nadir", "hot_bb", "cold_bb", "home")
```
Comment (dalonsoa) [inline tests/test_stepper_motor.py:6]: You can import this from the stepper motor module.

File: tests/test_stepper_motor.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+def test_constructor() -> None:
+    """Check that the constructor only accepts values in permissible range."""
+    with pytest.raises(ValueError):
+        DummyStepperMotor(0)
+
+    with pytest.raises(ValueError):
+        DummyStepperMotor(-1)
+
+    # Should work
+    DummyStepperMotor(len(PRESETS))
```
Comment (dalonsoa) [inline tests/test_stepper_motor.py:18]:
```suggestion
    for num in (-1, 0):
        with pytest.raises(ValueError):
            DummyStepperMotor(num)

    for num in (len(PRESETS), len(PRESETS)+10):
        DummyStepperMotor(num)
```

File: tests/test_stepper_motor.py [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+
+def test_move_to_number() -> None:
```
Comment (dalonsoa) [inline tests/test_stepper_motor.py:21]: I think that these tests that check that some values work well but others are not valid call for `pytest.parametrize` or even `hypothesis` to be really thorough a check edge cases.

File: finesse/hardware/__init__.py [hunk 8]
```diff
@@ -0,0 +1,5 @@
+"""This module contains code for interfacing with different hardware devices."""
+from .dummy_stepper_motor import DummyStepperMotor
+
+# TODO: Replace with a real stepper motor device
+stepper = DummyStepperMotor(3600)
```
Comment (dalonsoa) [inline finesse/hardware/__init__.py:5]: I don't think we want to create a global stepper here. We need to think about this, but considering how many devices we have, I think we should consider to setup a sort of `DeviceManager` living at the same level as the MainWindow in charge of starting them, dealing with their Errors, etc. For now, it is ok, but something to bear in mind.

**Review by dalonsoa — CHANGES_REQUESTED** [review]
There are a couple of suggestions both in the font and back end, but otherwise looks ok. I bet the real device will be more challenging...

File: finesse/gui/stepper_motor_view.py (hunk 2 above)
Comment (alexdewar) [reply finesse/gui/stepper_motor_view.py]: Ahh that's neat! I managed to rewrite my "make selected buttons blue" logic using a `QButtonGroup`, which is much cleaner.

File: tests/test_stepper_motor.py (hunk 5 above)
Comment (alexdewar) [reply tests/test_stepper_motor.py]: Good idea. I've moved it into `config.py` now so that we can also use it from the GUI side and not have to hardcode these values everywhere.

File: tests/test_stepper_motor.py (hunk 7 above)
Comment (alexdewar) [reply tests/test_stepper_motor.py]: Ah I've never used either of those, but it seems a sensible approach. I've ported over the tests to use `pytest.parametrize` now.

File: finesse/hardware/__init__.py (hunk 8 above)
Comment (alexdewar) [reply finesse/hardware/__init__.py:5]: Yep, I agree. Once the rest of the code takes shape it'll become a bit more obvious how to approach this.

**Comment by alexdewar** [comment]
I think it's ready for re-review now for when you're back.

In addition to the changes you suggested:

- I also rebased the branch onto `main`
- I swapped the `QGridLayout` in `MainWindow` for two separate columns of widgets, as `StepperMotorControl` was getting vertically stretched

File: finesse/hardware/dummy_stepper_motor.py (hunk 4 above)
Comment (alexdewar) [reply finesse/hardware/dummy_stepper_motor.py:49]: Oops! Yes we do

**Comment by alexdewar** [comment]
Ping @dalonsoa. Not urgent but if you have a chance to re-review this that'd be great.

It'd also be good to get your input @dc2917 while you're working on FINESSE.

**Review by dc2917 — APPROVED** [review]
Looks good to me, I can't see any issues here

File: finesse/gui/stepper_motor_view.py [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+        layout.addWidget(self.angle, 1, 2)
+        layout.addWidget(self.goto, 1, 3)
+
+        self.setLayout(layout)
+
+    def _add_checkable_button(self, name: str) -> QPushButton:
+        """Add a selectable button to button_group."""
+        btn = QPushButton(name)
+        btn.setCheckable(True)
+        btn.setStyleSheet("QPushButton::checked{background: blue; color: white;}")
```
Comment (dalonsoa) [inline finesse/gui/stepper_motor_view.py:50]: Unless you fancy specifically this style, buttons in a group change colour when they are pushed - at least in Mac; I would expect the same in others. See video. So I think this line can be removed.


https://user-images.githubusercontent.com/6095790/205260718-a79fd9c1-008e-48f1-a0b0-7e7ace8a1f28.mov

File: finesse/hardware/dummy_stepper_motor.py [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+        """
+        if isinstance(target, str):
+            self.current_step = DummyStepperMotor.get_preset_step(target)
+        else:
+            step = round(self.steps_per_rotation * target / 360.0)
+            if step < 0 or step >= self.steps_per_rotation:
+                raise ValueError("step number is out of range")
+            self.current_step = step
+
+        logging.info(f"Moving stepper motor to step {self.current_step}")
```
Comment (dalonsoa) [inline finesse/hardware/dummy_stepper_motor.py:52]: I was very confused by the log results, when I was selecting 1, clicking GOTO and having 10 printed in the terminal. I think that in the GUI there should be a clear indication that the number you are putting there is the angle in degrees and having the log showing:

```output
Moving stepper motor to step 10 (= 1º)
```

Or something like that.

File: finesse/gui/stepper_motor_view.py [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+        for i, preset in enumerate(ANGLE_PRESETS):
+            btn = self._add_checkable_button(preset.upper())
+            self.button_group.addButton(btn)
+
+            row, col = divmod(i, BUTTONS_PER_ROW)
+            layout.addWidget(btn, row, col)
+
+        # We also have a way for users to move the mirror to an angle of their choice
+        self.angle = QSpinBox()
+        self.angle.setMaximum(359)
```
Comment (dalonsoa) [inline finesse/gui/stepper_motor_view.py:38]: This makes the angular resolution equal to 1º. Is that good enough for the application?

File: tests/test_stepper_motor.py [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+def test_constructor_good(steps: int) -> None:
+    """Check valid arguments to constructor."""
+    DummyStepperMotor(steps)
+
+
+@pytest.mark.parametrize("steps", (-1000000, -1000, *range(-1, len(ANGLE_PRESETS))))
+def test_constructor_bad(steps: int) -> None:
+    """Check invalid arguments to constructor."""
+    with pytest.raises(ValueError):
+        DummyStepperMotor(steps)
```
Comment (dalonsoa) [inline tests/test_stepper_motor.py:20]: The normal use when using parametric tests is that you provide as input to the test the input parameters and the expected outputs/outcomes. So, these two functions could be written in a more concise manner as (see [VirtualRainforest](https://github.com/ImperialCollegeLondon/virtual_rainforest/blob/89407edba9c707f33281dd9051fecbee7ec716dd/tests/test_models.py#L43) for a more complex example):

```python
from contextlib import nullcontext as does_not_raise


@pytest.mark.parametrize(
    "steps,raises",
    [
        [
            num,
            pytest.raises(ValueError) if num < len(ANGLE_PRESETS) else does_not_raise(),
        ]
        for num in range(-5, len(ANGLE_PRESETS) + 5)
    ],
)
def test_constructor(steps: int, raises: Any) -> None:
    """Check invalid arguments to constructor."""
    with raises:
        DummyStepperMotor(steps)
```

Another option is to put the logic to decide on the outcome inside the test:

```python
from contextlib import nullcontext as does_not_raise


@pytest.mark.parametrize(
    "steps",
    list(range(-5, len(ANGLE_PRESETS) + 5)),
)
def test_constructor(steps: int) -> None:
    """Check invalid arguments to constructor."""
    raises = (
        pytest.raises(ValueError) if steps < len(ANGLE_PRESETS) else does_not_raise()
    )
    with raises:
        DummyStepperMotor(steps)
```

Assuming `mypy` does not complain much, I think this approach is more concise than splitting into good and bad tests.

**Review by dalonsoa — APPROVED** [review]
I have a few suggestions, but otherwise it looks good, so I'll approve it.

File: finesse/gui/stepper_motor_view.py (hunk 9 above)
Comment (alexdewar) [reply finesse/gui/stepper_motor_view.py:50]: Good point. I went with blue because that was the colour used in the old GUI, but I think the default style is probably fine.

For me, the buttons just appear as pressed in by default, not blue, but I guess that's fine too.

File: finesse/gui/stepper_motor_view.py (hunk 11 above)
Comment (alexdewar) [reply finesse/gui/stepper_motor_view.py:38]: It should be. I think that's what the old program has.

File: tests/test_stepper_motor.py (hunk 12 above)
Comment (alexdewar) [reply tests/test_stepper_motor.py:20]: Oh that's really cool. Thanks!
